app/controllers/course/survey/responses_controller.rb

```diff
@@ -24,20 +24,36 @@ def create
   end
 
   def show
-    render 'course/survey/surveys/index'
+    respond_to do |format|
+      format.html { render 'course/survey/surveys/index' }
+      format.json { render_response_json }
+    end
   end
 
   def edit
-    @response.build_missing_answers_and_options
-    if @response.save
-      render_response_json
-    else
-      head :internal_server_error
+    raise CanCan::AccessDenied if cannot?(:submit, @response) && cannot?(:modify, @response)
```

Comment:
Cannot just `authorize! :modify` since we can have the case where:
- Survey expired
- `allow_modify_after_submit` is false
- `allow_response_after_end` is true

So students who have not yet submitted can submit but not modify. If we only check `:modify`, students will not be able to access the `edit` page to submit in this case.

Cannot just `authorize! :submit` since, as discussed, we have the (unusual) case where:
- Survey expired
- `allow_modify_after_submit` is true
- `allow_response_after_end` is false

In this case, no students will be able to submit, but those who have submitted can modify.